src/tracing/instrumentation/mssql.js

```diff
@@ -0,0 +1,227 @@
+      }
+      return instrumentedFunction(this, originalFunction, originalArgs);
+    }
+    return originalFunction.apply(this, arguments);
+  };
+}
+
+
+function instrumentedRequestMethod(ctx, originalFunction, originalArgs) {
+  return instrumentedMethod(
+    ctx,
+    originalFunction,
+    originalArgs,
+    instrumentedRequestMethod,
+    function(args) {
+      return args[0];
+    }
+  );
+}
+
+
+function instrumentedBulk(ctx, originalFunction, originalArgs) {
+  return instrumentedMethod(
+    ctx,
+    originalFunction,
+    originalArgs,
+    instrumentedBulk,
+    function() {
+      return 'MSSQL bulk operation';
+    }
+  );
+}
+
+
+function instrumentedMethod(ctx, originalFunction, originalArgs, stackTraceRef, commandProvider) {
+  var parentSpan = cls.getCurrentSpan();
+
+  if (cls.isExitSpan(parentSpan)) {
+    return originalFunction.apply(ctx, originalArgs);
+  }
+
+  var connectionParameters = findConnectionParameters(ctx);
+  var command = commandProvider(originalArgs);
+  return cls.ns.runAndReturn(function() {
+    var span = cls.startSpan('mssql');
+    span.stack = tracingUtil.getStackTrace(stackTraceRef);
+    span.data = {
+      mssql: {
+        stmt: tracingUtil.shortenDatabaseStatement(command),
+        host: connectionParameters.host,
+        port: connectionParameters.port,
+        user: connectionParameters.user,
+        db: connectionParameters.db
+      }
+    };
+
+    var originalCallback;
+    if (originalArgs.length >= 2 && typeof originalArgs[1] === 'function') {
+      originalCallback = cls.ns.bind(originalArgs[1]);
+    }
+
+    if (originalCallback) {
+      // original call had a callback argument, replace it with our wrapper
+      var wrappedCallback = function(error) {
+        finishSpan(error, span);
+        return cls.ns.bind(originalCallback).apply(this, arguments);
+      };
+      originalArgs[1] = cls.ns.bind(wrappedCallback);
+    }
+
+    var promise = originalFunction.apply(ctx, originalArgs);
+    if (typeof promise.then === 'function') {
+      promise.then(function(value) {
+        finishSpan(null, span);
+        return value;
+      })
+      .catch(function(error) {
+        finishSpan(error, span);
+        return error;
+      });
+    }
+    return promise;
+  });
+}
+
+
+function instrumentPreparedStatement(PreparedStatement) {
+  shimmer.wrap(PreparedStatement.prototype, 'prepare', shimPrepare);
+  shimmer.wrap(PreparedStatement.prototype, 'execute', shimMethod.bind(null, instrumentedExecute));
+}
+
+
+function shimPrepare(originalFunction) {
+  return function() {
+    if (isActive && cls.isTracing()) {
+      var originalArgs = new Array(arguments.length);
+      for (var i = 0; i < arguments.length; i++) {
+        originalArgs[i] = arguments[i];
+      }
+      if (originalArgs.length >= 2 && typeof originalArgs[1] === 'function') {
+        originalArgs[1] = cls.ns.bind(originalArgs[1]);
+      }
+      cls.ns.set('com.instana.mssql.stmt', originalArgs[0]);
```

Comment:
Currently every `cls.ns.set` is added to a span for cleanup purposes. Also, when I understand correctly, then there is no guarantee that a prepared statement is executed within the same async context as the one used to prepare it. So as far as I can tell, this won't work?

As an alternative, couldn't we register a property `__instanaStatement` or similar on the `PreparedStatement` so that we have access to this later on? That way we also don't need to worry about cls cleanup.

Reply:
Hm, I think the `cls.ns.bind(originalArgs[1])` makes it work, even if the context is switched. I'll look into the `__instanaStatement` approach though, I agree that it would be nicer.